### PR TITLE
Added `"any"` mode to `register_multi_grad_hook`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* #117867
* #117861
* #117856
* #117852
* #117817
* #117814
* #117776
* __->__ #117984
* #118186
* #117994

This is a re-open of https://github.com/pytorch/pytorch/pull/115628/. This PR adds an `"any"` option to `register_multi_grad_hook` that runs the hook when the gradient of _any_ of the input tensors is computed. The existing functionality is folded under the default `"all"` mode.

The multi-threaded test case is based on the existing one for `register_multi_grad_hook`. I would appreciate a closer look on that. ~~I am not sure about the hook signature (i.e. why we see two gradients in the hook that runs instead of just one, as [`register_hook`](https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html) docs suggest).~~ It was because I was iterating over the 2 elements in the single tensor 😢 .

I did not update the `notes/autograd.rst`, which currently has a [blurb](https://pytorch.org/docs/stable/notes/autograd.html#special-hooks) on `register_multi_grad_hook`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -458,11 +444,63 @@ def __getstate__(self):
...
+                count[id] += 1
+
+                if count[id] == nb_calls:
+                    fn(buffer[id])
+                    del count[id]
+                    del buffer[id]
+
+            return inner_hook
+
+        handles: Tuple[RemovableHandle] = tuple(
```
Comment (awgu) [inline torch/autograd/graph.py:479]: For the existing implementation, the change is just code motion (indent) plus changing this to a `tuple` comprehension (for easier typing with the new code).

**Comment by albanD** [comment]
@soulitzer I'll let you review this one!

File: test/test_autograd.py [hunk 2]
```diff
@@ -8597,6 +8599,67 @@ def backward(ctx, gO):
...
+                nonlocal hook_id
+                z.register_hook(partial(hook, hook_id))
+                hook_id += 1
+                any_hook_handles.append(
+                    torch.autograd.graph.register_multi_grad_hook(
+                        out, partial(hook, hook_id), mode="any"
+                    )
+                )
+                hook_id += 1
+                return out[0] + out[1]
```
Comment (soulitzer) [inline test/test_autograd.py:8623]: Did you mean to return out? I see that you sum it again later when using this.

**Review by soulitzer — COMMENTED** [review]
Thanks! Looks great, just some minor comments

File: test/test_autograd.py (hunk 2 above)
Comment (awgu) [reply test/test_autograd.py:8623]: Good catch! Thanks!

File: test/test_autograd.py [hunk 3]
```diff
@@ -11209,6 +11272,73 @@ def backward_retain_graph(out, t2, t3):
...
+        bw_count = [0]
+
+        class Func(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                return x
+
+            @staticmethod
+            def backward(ctx, gO):
+                bw_count[0] += 1
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/test_autograd.py]: I added a lock to the existing and new tests.

File: torch/autograd/graph.py [hunk 4]
```diff
@@ -370,11 +384,18 @@ def disable_saved_tensors_hooks(error_message):
 
 def register_multi_grad_hook(
     tensors: Sequence[torch.Tensor],
-    fn: Callable[[Sequence[Optional[torch.Tensor]]], None],
+    fn: Union[
```
Comment (awgu) [inline torch/autograd/graph.py:387]: @soulitzer The typing is kind of cooked because `mode="all"` and `mode="any"` accept different `fn` signatures.

File: test/test_autograd.py [hunk 5]
```diff
@@ -11209,6 +11285,75 @@ def backward_retain_graph(out, t2, t3):
...
+        # even if those executions occur concurrently on different threads
+        t1 = torch.rand(2, requires_grad=True)
+        t2 = torch.rand(2, requires_grad=True)
+        t3 = torch.rand(2, requires_grad=True)
+        t4 = torch.rand(2, requires_grad=True)
+
+        res = None
+        count = [0]
+
+        def hook(grads):
```
Comment (soulitzer) [inline test/test_autograd.py:11300]: This only needs to take a single tensor

File: test/test_autograd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply test/test_autograd.py]: Thanks!
Btw we probably need them for the other global state shared across threads: res and counts (I'm sorry these existing tests are broken lol).

File: torch/autograd/graph.py (hunk 4 above)
Comment (soulitzer) [reply torch/autograd/graph.py:387]: Ah welp, I guess we will have to live with that lol 

This makes me wonder if an extra sentence in the docs would be slightly clear:
after ``a tensor in :attr:`tensors` has been computed....``
maybe add something like `The hook will be called with the single gradient that has been computed`

Comment (awgu) [reply torch/autograd/graph.py:387]: I added a sentence 😄

File: test/test_autograd.py (hunk 5 above)
Comment (awgu) [reply test/test_autograd.py]: Fixed!

File: test/test_autograd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awgu) [reply test/test_autograd.py]: I should have added locks for the global state.

Note to self: It looks like `threading.Lock` should get released appropriately when a thread raises an exception (https://stackoverflow.com/questions/39050277/in-python-does-lock-get-automatically-released-when-an-exception-happens), so locking over a region with `self.assert...` should be fine.

**Review by soulitzer — APPROVED** [review]
LGTM!

**Comment by awgu** [comment]
@soulitzer It looks like the BC-linter is complaining about the function signature change:
```
##[warning]Function register_multi_grad_hook: fn changed from Callable[<ast.List object at 0x7f12125c1150>, None] to Union[Callable[<ast.List object at 0x7f1212601a50>, None], Callable[<ast.List object at 0x7f1212601840>, None]]
```
What to do 🤔 

What is the consequence of this type change? We are making the typing less strict -- would this cause any typing issues for existing code? I did not think so, but since the BC-linter is complaining, I am not sure.

**Comment by awgu** [comment]
Regarding BC-linter:
- The main concern is that changing from type `A` to `Union[A, B]` could break TorchScripted models.
- However, TorchScript does not support backward hooks: https://github.com/pytorch/pytorch/issues/34329

In that case, we should be okay to suppress the BC-lint and land.

**Comment by awgu** [comment]
Another comment is that `register_multi_grad_hook` is not widely adopted, so the blast radius here is small. E.g., the only adopter in PyTorch code is FSDP. We can re-triage if something does break from this, but it should be unlikely.
